ui/src/app/utils/stores.tsx

```diff
@@ -0,0 +1,23 @@
+import {useEffect, useState} from 'react';
+import {atom, Atom} from './subscribable';
+
+interface RouteDataStore {
+  title?: string;
+}
+
+export const routeDataStore = atom <RouteDataStore>({});
+
+export interface AuthStore {
+  authLoaded: boolean;
+  isSignedIn: boolean;
+}
+
+export const authStore = atom<AuthStore>({authLoaded: false, isSignedIn: false});
+
+export function useStore<T>(theStore: Atom<T>) {
+  const [value, setValue] = useState(theStore.get());
+  useEffect(() => {
+    return theStore.subscribe(v => setValue(v)).unsubscribe;
+  }, [theStore, setValue]);
```

Comment:
I looked at the useEffect docs. This line is still does not compute.

`theStore` is not a property or state, AFAICT. What does it mean to filter `useEffect` to only run when it "changes"? setValue might make some sense(?); my best guess is that it basically daisy-chains the above subscription setup to itself.

Can you add a comment, or if there's a more general concept I'm missing - point me to the right docs?

Reply:
Thanks, this makes more sense after reading the React Hooks docs through.

Let me check my understanding:

- `theStore` in all current use cases is not a prop or state, but this function should treat all incoming parameters as if they might be props
- Presumably the `useEffect` filter works on non-prop/state values. I'm imagining in this specific case where `theStore` is not a prop, `[theStore]` is equivalent to `[]` ?
- Accordingly, we expect this `useEffect()` to run once on initial mount, then never run again

Reply:
Yes, useEffect will work on any value. There could be a case where there is a new value derived via a computation between state and props and only want to run the effect in that case. In the vast majority of cases this will run once. But in the event there are multiple stores of the same type, and a component swaps out the store in the hook, the proper cleanup (unsubscribe) and setup (subscribe) will take place for the newly designated store.